### ISSUE TITLE
Error building graph with `@tf.function` decorator for recursive function with strided slices

### ISSUE THREAD

**Issue by ValadeAurelien** [tensorflow/tensorflow#45279]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux version 5.9.8-arch1-1
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): pip install tensorflow --user 
- TensorFlow version (use command below): v2.3.0-54-gfcc4b966f1 2.3.1
- Python version: python 3.8.6-1
- Bazel version (if compiling from source): 
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: None
- GPU model and memory: None

**Describe the current behavior**

I coded my own 3 dimensional FFT with TensorFlow, translated from the algorithm `In[4]` of [https://jakevdp.github.io/blog/2013/08/28/understanding-the-fft/](https://jakevdp.github.io/blog/2013/08/28/understanding-the-fft/).

I know there is already an FFT coded in TensorFlow, but it does not have a gradient implemented, which I need.

The code works without the decorators `@tf.function` and gives results coherent with `numpy`. However it crashes while building the graph if they are present, invoking an error on dimensions (ill computed in the case of graph computation). I have not been able to trace the problem to its core. 

**Describe the expected behavior**

I would expect the code to work with decorators, however I must admit that the use of strided slices in a recursive function might be not be compatible with automatic graph building (lines 30 and 31).

**Standalone code to reproduce the issue**

```
#!/usr/bin/env python3

import numpy as np
import tensorflow as tf

dtype = 'float32'
cdtype = 'complex64'
c_pi = tf.constant(np.pi, cdtype)

# Leaves of the FFT algorithm : DFT of 32
# Just building a 32 x 32 complex matrix
tf_Nmin = 32
tf_rg = tf.range(tf_Nmin)
tf_k = tf.broadcast_to(tf_rg, (tf_Nmin, tf_Nmin))
tf_n = tf.transpose(tf_k)
tf_M = tf.exp(-2j * c_pi * tf.cast(tf_k * tf_n / tf_Nmin, cdtype))


# Matrix product between M and x with shape = [32, 128, 128] 
# Works without the decorator
# Fails with the decorator, invoking that the shapes are [32,32] and [128,16384]. 
@tf.function
def _tf_FFT0_leaf(x):
    return (tf.tensordot(tf_M, x, axes=[[1], [0]]))


# Nodes of the FFT algorithm
@tf.function
def _tf_FFT0_node(x, N):
    N2 = N // 2
    XE = tf_FFT0(x[::2, :, :], N2)
    XO = tf_FFT0(x[1::2, :, :], N2)
    fac = tf.exp(-2j * c_pi * tf.cast(tf.range(N2) / N, cdtype))
    fac = tf.transpose(tf.broadcast_to(fac, [tf_N, tf_N, N2]), perm=(2, 0, 1))
    fXO = fac * XO
    return (tf.concat([XE + fXO, XE - fXO], axis=0))


# Recursive function of the FFT algorithm
@tf.function
def tf_FFT0(x, N):
    return (tf.cond(N <= tf_Nmin, lambda: _tf_FFT0_leaf(x),
                    lambda: _tf_FFT0_node(x, N)))


# Run the code
nx = 128
x = np.random.random((nx, nx, nx))
tf_x = tf.cast(x, cdtype)
tf_N = tf.constant(x.shape[0], 'int32')

y = tf_FFT0(tf_x, tf_N)
z = np.fft.fft(x, axis=0)
print()
print('DIFF WITH NUMPY', np.mean(np.abs(y - z) / np.abs(z)))
```
**Other info / logs** 

Output **without** decorator:
```
2020-11-30 19:51:41.195663: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'libcudart.so.10.1'; dlerror: libcudart.so.10.1: cannot open shared object file: No such file or directory
2020-11-30 19:51:41.195690: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
2020-11-30 19:51:42.046458: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory
2020-11-30 19:51:42.046475: W tensorflow/stream_executor/cuda/cuda_driver.cc:312] failed call to cuInit: UNKNOWN ERROR (303)
2020-11-30 19:51:42.046485: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (lyonovae21.in2p3.fr): /proc/driver/nvidia/version does not exist
2020-11-30 19:51:42.046697: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN)to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2020-11-30 19:51:42.066848: I tensorflow/core/platform/profile_utils/cpu_utils.cc:104] CPU Frequency: 1900610000 Hz
2020-11-30 19:51:42.067237: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x5590a7465cf0 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2020-11-30 19:51:42.067308: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version

DIFF WITH NUMPY 4.81155889723075e-06
```

Output **with** decorator:
```
2020-11-30 20:13:22.743043: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'libcudart.so.10.1'; dlerror: libcudart.so.10.1: cannot open shared object file: No such file or directory
2020-11-30 20:13:22.743073: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
2020-11-30 20:13:23.540001: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory
2020-11-30 20:13:23.540022: W tensorflow/stream_executor/cuda/cuda_driver.cc:312] failed call to cuInit: UNKNOWN ERROR (303)
2020-11-30 20:13:23.540035: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (lyonovae21.in2p3.fr): /proc/driver/nvidia/version does not exist
2020-11-30 20:13:23.540281: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN)to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2020-11-30 20:13:23.560290: I tensorflow/core/platform/profile_utils/cpu_utils.cc:104] CPU Frequency: 1900610000 Hz
2020-11-30 20:13:23.560760: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x5602b27b1590 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2020-11-30 20:13:23.560786: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
Traceback (most recent call last):
  File "crash_tf.py", line 54, in <module>
    y = tf_FFT0(tf_x, tf_N)
  File "/home/valade/.local/lib/python3.8/site-packages/tensorflow/python/eager/def_function.py", line 780, in __call__
    result = self._call(*args, **kwds)
  File "/home/valade/.local/lib/python3.8/site-packages/tensorflow/python/eager/def_function.py", line 823, in _call
    self._initialize(args, kwds, add_initializers_to=initializers)
  File "/home/valade/.local/lib/python3.8/site-packages/tensorflow/python/eager/def_function.py", line 696, in _initialize
    self._stateful_fn._get_concrete_function_internal_garbage_collected(  # pylint: disable=protected-access
  File "/home/valade/.local/lib/python3.8/site-packages/tensorflow/python/eager/function.py", line 2855, in _get_concrete_function_internal_garbage_collected
    graph_function, _, _ = self._maybe_define_function(args, kwargs)
  File "/home/valade/.local/lib/python3.8/site-packages/tensorflow/python/eager/function.py", line 3213, in _maybe_define_function
    graph_function = self._create_graph_function(args, kwargs)
  File "/home/valade/.local/lib/python3.8/site-packages/tensorflow/python/eager/function.py", line 3065, in _create_graph_function
    func_graph_module.func_graph_from_py_func(
  File "/home/valade/.local/lib/python3.8/site-packages/tensorflow/python/framework/func_graph.py", line 986, in func_graph_from_py_func
    func_outputs = python_func(*func_args, **func_kwargs)
  File "/home/valade/.local/lib/python3.8/site-packages/tensorflow/python/eager/def_function.py", line 600, in wrapped_fn
    return weak_wrapped_fn().__wrapped__(*args, **kwds)
  File "/home/valade/.local/lib/python3.8/site-packages/tensorflow/python/framework/func_graph.py", line 973, in wrapper
    raise e.ag_error_metadata.to_exception(e)
ValueError: in user code:

    crash_tf.py:43 tf_FFT0  *
        return (tf.cond(N <= tf_Nmin, lambda: _tf_FFT0_leaf(x),
    crash_tf.py:21 _tf_FFT0_leaf  *
        return (tf.tensordot(tf_M, x, axes=[[1], [0]]))
    /home/valade/.local/lib/python3.8/site-packages/tensorflow/python/util/dispatch.py:201 wrapper  **
        return target(*args, **kwargs)
    /home/valade/.local/lib/python3.8/site-packages/tensorflow/python/ops/math_ops.py:4518 tensordot
        ab_matmul = matmul(a_reshape, b_reshape)
    /home/valade/.local/lib/python3.8/site-packages/tensorflow/python/util/dispatch.py:201 wrapper
        return target(*args, **kwargs)
    /home/valade/.local/lib/python3.8/site-packages/tensorflow/python/ops/math_ops.py:3253 matmul
        return gen_math_ops.mat_mul(
    /home/valade/.local/lib/python3.8/site-packages/tensorflow/python/ops/gen_math_ops.py:5640 mat_mul
        _, _, _op, _outputs = _op_def_library._apply_op_helper(
    /home/valade/.local/lib/python3.8/site-packages/tensorflow/python/framework/op_def_library.py:742 _apply_op_helper
        op = g._create_op_internal(op_type_name, inputs, dtypes=None,
    /home/valade/.local/lib/python3.8/site-packages/tensorflow/python/framework/func_graph.py:591 _create_op_internal
        return super(FuncGraph, self)._create_op_internal(  # pylint: disable=protected-access
    /home/valade/.local/lib/python3.8/site-packages/tensorflow/python/framework/ops.py:3477 _create_op_internal
        ret = Operation(
    /home/valade/.local/lib/python3.8/site-packages/tensorflow/python/framework/ops.py:1974 __init__
        self._c_op = _create_c_op(self._graph, node_def, inputs,
    /home/valade/.local/lib/python3.8/site-packages/tensorflow/python/framework/ops.py:1815 _create_c_op
        raise ValueError(str(e))

    ValueError: Dimensions must be equal, but are 32 and 128 for '{{node Tensordot/MatMul}} = MatMul[T=DT_COMPLEX64, transpose_a=false, transpose_b=false](Tensordot/a, Tensordot/Reshape)' with input shapes: [32,32], [128,16384].
```

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with [TF v2.3](https://colab.research.google.com/gist/amahendrakar/004c7c6d9d15ef277811c589beadf53f/45279.ipynb) and [TF-nightly](https://colab.research.google.com/gist/amahendrakar/cfd457b9fd70a2e94bbe964a0bf3265a/45279-tf-nightly.ipynb). Please find the attached gist. Thanks!

**Comment by mdanatg** [comment]
Unfortunately, you're right, `tf.function` doesn't support recursion at the moment. It is possible to implement in theory, and it's a feature that's being asked now and then.

The error message on the other hand is a confusing side effect of tracing combined with static shape verification: when tracing `tf_FFT0`, it is traced with the most specific shape of its input, which happens to be `(128, 128, 128)`. Then the tf.cond traces both its branches, including `_tf_FFT0_leaf`. But when tracing that, we get a `matmul` which checks the current shape of x (`(128, 128, 128)`) and `tf_M` and complains that they're mismatched. The tracer has no idea that `_tf_FFT0_leaf` will never be called with this shape. That issue is resolved by telling `tf.function` to assume that the shape is dynamic so it should skip this static verification. We can do that like so:

```
@tf.function(
    input_signature=[
                     tf.TensorSpec(shape=[None, nx, nx], dtype=cdtype),
                     tf.TensorSpec(shape=[], dtype=tf.int32)])
def tf_FFT0(x, N):
  ...
```

(this is something worth doing to avoid excessive retraces as well)

Alas, this only explains the strange error message, and doesn't resolve the main problem which is that tf.function doesn't support recursion.

It's always possible to rewrite a recursive algorithm in a non-recursive fashion, using a few while loops and `tf.TensorArray` as a stack, see the example below.
But first I should ask whether the ops in [tf.signal](https://www.tensorflow.org/api_docs/python/tf/signal) have the FFT implementations that you need?

Anyway here's the non-recursive example. It's a bit long because working with `TensorArray` is very verbose. But we can use autograph to write the control flow as normal Python:

```
@tf.function(
    input_signature=[
                     tf.TensorSpec(shape=[None, nx, nx], dtype=cdtype),
                     tf.TensorSpec(shape=[], dtype=tf.int32)])
def tf_FFT0(x, N):
    s_x = tf.TensorArray(cdtype, size=0, dynamic_size=True, element_shape=[None, nx, nx])  # Set up a stack for x.
    s_N = tf.TensorArray(tf.int32, size=0, dynamic_size=True)  # Set up a stack for N.
    s_ret = tf.TensorArray(cdtype, size=0, dynamic_size=True, infer_shape=False)  # Set up a stack for the results.

    # Set up a stack for the state needed to simulate the recursion.
    # Node states:
    #  0 - entry, need to calculate tf_FFT0(x[::2, :, :], N2)
    #  1 - need to calculate tf_FFT0(x[1::2, :, :], N2)
    #  2 - both operands done, compute the final result
    s_node = tf.TensorArray(tf.int32, size=0, dynamic_size=True)

    # TensorArray doesn't support deletion, so keep a stack pointer separately.
    level = 0
    s_x = s_x.write(level, x)
    s_N = s_N.write(level, N)
    s_node = s_node.write(level, 0)

    level = tf.constant(0)  # For autograph - we need a TF while_loop.
    while level >= 0:
      N = s_N.read(level)
      x = s_x.read(level)
      if N <= tf_Nmin:
        # leaf
        # No recursion. Put results on stack and go back.
        s_ret = s_ret.write(level, tf.tensordot(tf_M, x, axes=[[1], [0]]))
        level -= 1
      else:
        # node
        # Simulate recursion.
        state = s_node.read(level)
        N2 = N // 2
        if state == 0:
          # Nothing calculated yet.
          s_node = s_node.write(level, 1)

          # "Call" for first operand.
          level += 1
          s_x = s_x.write(level, x[::2, :, :])
          s_N = s_N.write(level, N2)
          s_node = s_node.write(level, 0)

        elif state == 1:
          # First operand done.
          s_node = s_node.write(level, 2)

          # Copy result for first operand to current level.
          s_ret = s_ret.write(level, s_ret.read(level + 1))

          # "Call" for second operand.
          level += 1
          s_x = s_x.write(level, x[1::2, :, :])
          s_N = s_N.write(level, N2)
          s_node = s_node.write(level, 0)

        else:
          # Both operands done. Current level has first operand, next level has the second.
          XE = s_ret.read(level)
          XO = s_ret.read(level + 1)
          fac = tf.exp(-2j * c_pi * tf.cast(tf.range(N2) / N, cdtype))
          fac = tf.transpose(tf.broadcast_to(fac, [tf_N, tf_N, N2]), perm=(2, 0, 1))
          fXO = fac * XO

          # All done. Put final result on stack and go back.
          s_ret = s_ret.write(level, tf.concat([XE + fXO, XE - fXO], axis=0))
          level -= 1

    # Final results should be at the bottom of the stack.
    return s_ret.read(0)
```

**Comment by ValadeAurelien** [comment]
Thanks a lot for your very detailed answer! 

I am surprised that it is possible for the algorithm to draw a graph from a python function containing a `while` statement! Does it impact the performances? Is `tensorflow` going to be able to compute the gradient of that function ?

I am using the HMC method implemented in `tensorflow_probabilities` but in order to compute my likelihood, I need to perform a 3D-FT on the parameters and interpolate the resulting field on a few dozen of thousands of points... However, letting the interpolation part aside, the `tf.signal.irfft3d` does not have a gradient implemented, which led to a error while trying to run the HMC. I do not have a example code to hand right now, but the error simply read in substance as `tf.signal.irfft3d does not have a gradient method implemented`. I then decided to compute my own, using as `tf.function` to efficiently build the graph.

**Comment by mdanatg** [comment]
The graph is drawn by autograph which uses source code transformation. Simply put, it turns the `while` statement into a `tf.while_loop` call; the transformation can be done mechanically. And the result is differentiable so gradients should work, too.

Performance-wise, it's identical to writing `tf.cond` and `tf.while_loop` by hand (but less annoying). A word of caution that `tf.cond` and `tf.while_loop` are not terribly fast themselves. They do become much faster by setting the `experimental_compile=True` arg of `tf.function`, although that has some limitations, and I think it ran into a bug that we'll need to fix before we can use it. Hopefully we should be able to try it in a few weeks.

**Comment by ValadeAurelien** [comment]
Thank you again for your answer. 
Do you thus confirm there is no simpler work-around to get the gradient of the (inverse) Fourrier Transform in `tensorflow` ?
Should I then close the issue?

**Comment by mdanatg** [comment]
Yes, I think the example I gave is about as simple as it can get, at the moment. We're considering supporting recursion in `tf.function`, but there are no concrete plans for it yet.
Technically one could also define a gradient kernel for `tf.signal.irfft3d`, but that is not trivial either and involves writing C++ code.
So I believe it's ok to close this issue, and I hope the alternate solution I gave is useful!
